rivertest/rivertest.go

```diff
@@ -535,3 +536,12 @@ func failure(t testingT, format string, a ...any) {
 func failureString(format string, a ...any) string {
 	return "\n    River assertion failure:\n    " + fmt.Sprintf(format, a...) + "\n"
 }
+
+// WorkContext returns a realistic context that can be used to test JobArgs.Work
+// implementations.
+//
+// In particual, adds a client to the context so that river.ClientFromContext is
+// usable in the test suite.
+func WorkContext[TTx any](ctx context.Context, client *river.Client[TTx]) context.Context {
+	return context.WithValue(ctx, rivercommon.ContextKeyClient{}, client)
+}
```

Comment:
Do you think this should also attempt to apply timeouts to the context or no? For example there's a default client-level job timeout, as well as the possibility of it being overridden by the worker's `Timeout(job)` method.

I did that with my internal job testing helper in the API, but since it's worker-specific maybe it makes more sense for the possible job-running helper instead of this one.

Reply:
I definitely thought about it, but I think it might be out of scope for this particular helper.

Once added, a timeout is hard to remove (so I'd generally bias against it), but more importantly here, we don't have the right information to calculate what a timeout should be — to know what your timeout is, you need access to the client default, the job args to pull an `InsertOpts` off of, and the job itself in case of a job-specific override.

I think we could add something like this with a more elaborate `rivertest.Work` function of some kind, but it'd be a bit more of a project. I tried to prototype this quickly, but it's easier said than done because you'd need some back channel to pull a timeout out of an injected client, and possibly a way of getting an executor from job args.

We could experiment with that path a little more, but I suspect it might be easier to just have users explicitly call their own `Work` functions and add their own timeouts at that time if they want it. We'd supply a basic context with `WorkContext`, but nothing else.

I know I always beat this point home, but I'll just float again that use of `context.WithTimeout` in tests can be considered a major anti-pattern because in case one fails, you get very little information about that where a timeout happened. Compare that to using `go test ./... -timeout 1m` and you get an exact stack trace of where a slow test was when it hits its maximum time allowance.

tl;dr I think we should probably leave timeouts of the equation for now, and possible consider them for more elaborate test functions down the line.

Reply:
> I know I always beat this point home, but I'll just float again that use of `context.WithTimeout` in tests can be considered a major anti-pattern because in case one fails, you get very little information about that where a timeout happened.

I agree generally but `WithTimeoutCause` does change the story here a little bit, particularly if the error is easily identifiable. Otherwise I agree with the rest of what you wrote 😄

Reply:
Related to my last, I checked and we do not actually use `WithTimeoutCause` anywhere—I think because we wanted to support Go 1.20 when we initially launched River. Maybe we should make use of that for River-controlled timeouts?

Reply:
Yeah, agreed that it'd probably make sense to replace all our timeouts with a `WithTimeoutCause` and then have a single River timeout value that people could reference. Not sure how often people really need to check what kind of timeout occurred, but I'm sure it'll be useful to somebody.

TBC, are you saying that we should still add `WithTimeoutCause` here too, or just more broadly use it elsewhere?

Reply:
Thinking about it more, I think the test helper is probably best _without_ applying its own timeouts. I don't think most people will want to be running unit tests with their real world timeouts applied anyway.

So yeah, more of a side comment, this is good to go 👍

Reply:
Perfect, thx.